<a href="https://colab.research.google.com/github/benny4421/MLOps/blob/main/MlopsHW1Q5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Question 5 - Process Time Simulation

simulated 1,000,000 times

The development path and the data pipeline run in parallel. Model Productionization starts only after both paths are finished.


In [6]:
import numpy as np
import pandas as pd



n = 1_000_000

In [2]:
sh = np.random.exponential(scale=10, size=n)

data_extract = np.random.normal(loc=15, scale=3, size=n)
dev_exploration = np.random.normal(loc=3, scale=0.5, size=n)
dev_modeling = np.random.poisson(lam=30, size=n)

data_pipeline = np.random.poisson(lam=100, size=n)

model_production = np.random.normal(loc=30, scale=5, size=n)
process_integration = np.random.normal(loc=50, scale=20, size=n)

In [3]:
dev_path = data_extract + dev_exploration + dev_modeling

total_time = (
    sh
    + np.maximum(dev_path, data_pipeline)
    + model_production
    + process_integration
)

print("Average:", np.mean(total_time))
print("Median:", np.median(total_time))
print("95% interval:", np.percentile(total_time, [2.5, 97.5]))

Average: 190.02111410181516
Median: 189.5498194916108
95% interval: [142.36210724 240.54022139]


The average total process time is about 190 days, and the median is about 189 days. The middle 95% of simulated process times is approximately 142 to 241 days.

In [4]:
steps = pd.DataFrame({
    "Step": [
        "SH Analysis",
        "Data Extract",
        "Dev Exploration",
        "Dev Modeling",
        "Data Pipeline",
        "Model Productionization",
        "Process Integration"
    ],
    "Mean": [
        np.mean(sh),
        np.mean(data_extract),
        np.mean(dev_exploration),
        np.mean(dev_modeling),
        np.mean(data_pipeline),
        np.mean(model_production),
        np.mean(process_integration)
    ],
    "Std Dev": [
        np.std(sh),
        np.std(data_extract),
        np.std(dev_exploration),
        np.std(dev_modeling),
        np.std(data_pipeline),
        np.std(model_production),
        np.std(process_integration)
    ]
})

steps

,Step,Mean,Std Dev
0,SH Analysis,10.011138,10.010365
1,Data Extract,14.994140,3.003292
2,Dev Exploration,3.000150,0.499995
3,Dev Modeling,30.008560,5.481296
4,Data Pipeline,100.007608,9.999428
5,Model Productionization,30.003963,5.000939
6,Process Integration,49.998397,20.004666


The Data Pipeline contributes the most to the total process time because it has the largest average duration. Process Integration contributes the most uncertainty because it has the largest standard deviation.

In [5]:
data_pipeline_new = np.random.poisson(lam=150, size=n)

total_time_new = (
    sh
    + np.maximum(dev_path, data_pipeline_new)
    + model_production
    + process_integration
)

delay = total_time_new - total_time

print("New average total time:", np.mean(total_time_new))
print("Average delay:", np.mean(delay))
print("95% interval for delay:", np.percentile(delay, [2.5, 97.5]))

New average total time: 240.01711356101347
Average delay: 49.99599945919829
95% interval for delay: [19. 81.]


When the average Data Pipeline time increases from 100 to 150 days, the overall process is delayed by about 50 days on average. The middle 95% of simulated delays is approximately 19 to 81 days.
